### DISCUSSION TITLE
Need help to improve an implementation

### DISCUSSION THREAD

**Discussion by raanam** [vkhorikov/CSharpFunctionalExtensions#534] [Q&A] [answered]
Presently I am using Functional Extension in following way

``` cs
Result<EntityA> a= await ResolveEntityAFromDb(request.AId, token);
if (a.IsFailure) return a.Error;

Result<EntityB> b = await ResolveEntityBFromDb(request.BId, token);
if (b.IsFailure) return b.Error;

EntityC c = Map(a.Result, b.Result, request);
await dbContext.SaveChanges();

return Result.Success;
```

I want to avoid having if(a.IsFailure), if (b.IsFailure)  checks and make it railway oriented. Any Ideas of how I can chain such operations and use results of all previous operations (ResolveEntityAFromDb, ResolveEntityBFromDb) in last step.

**Comment by hankovich** [comment, marked as answer]
Hi, you can use something like 

```c#
return ResolveEntityAFromDb(request.AId, token)
    .Bind(result1 => ResolveEntityBFromDb(request.BId, token).Map(result2 => (result1, result2))
    .Map(tuple => Map(tuple.result1, tuple.result2, request))
    .Tap(() => dbContext.SaveChanges());
```

**Reply by hankovich** [reply]
Please also take a look at https://github.com/vkhorikov/CSharpFunctionalExtensions/issues/535, once implemented it will fit your needs

**Reply by raanam** [reply]
Thanks for helping me out here. Definitely a use case for me as well.

I was able to use linq Query syntax to achieve what I needed

Like - following works but I cannot understand to be honest why it works, even though I like the syntax

`
var  result = from entityA in ResolveEntityAFromDb(request.AId, token) `<br>`
                    from entityB in ResolveEntityBFromDb(request.BId, token) `<br>`
                    select Map(entityA, entityB);
`

**Comment by vkhorikov** [comment]
Yeah, we can add the additional extension for sure.

I usually use a workaround like this:

```cs
Result result1 = ResolveEntityAFromDb(request.AId, token);

result1.Bind(_ => ResolveEntityBFromDb(request.BId, token)
  .Map(entityB => Map(result1.Value, entityB).
  ....
```

**Reply by NormanMann** [reply]
My repositories usually don't return Results but look like : 

public async Task<EntityA?> ResolveEntityAFromDB() {}

Would that mean that - now that we have BindZip - MapZip would make sense as well ?

**Reply by vkhorikov** [reply]
I guess so
